### ISSUE TITLE
Mouse pointer disappears while dragging windows

### ISSUE THREAD

**Issue by cplir-c** [swaywm/sway#8340]
### Please read the following before submitting:
- Please do NOT submit bug reports for questions. Ask questions on IRC at #sway on Libera Chat.
- Proprietary graphics drivers, including nvidia, are not supported. Please use the open source equivalents, such as nouveau, if you would like to use Sway.
- Please do NOT submit issues for information from the github wiki. The github wiki is community maintained and therefore may contain outdated information, scripts that don't work or obsolete workarounds.
  If you fix a script or find outdated information, don't hesitate to adjust the wiki page.

### Please fill out the following:
- **Sway Version:**
  - `swaymsg -t get_version` or `sway -v`
```sh
$ swaymsg -t get_version
sway version 1.9
$ pacman -Q sway
sway 1:1.9-5
```

- **Debug Log:**
  - Run `sway -d 2> ~/sway.log` from a TTY and upload it to a pastebin, such as gist.github.com.
  - This will record information about sway's activity. Please try to keep the reproduction as brief as possible and exit sway.
  - Attach the **full** file, do not truncate it.
  [file](https://gist.github.com/cplir-c/68411d75c54f59626be8c6306ed4234a)

- **Configuration File:**
  - Please try to produce with the default configuration.
  - If you cannot reproduce with the default configuration, please try to find the minimal configuration to reproduce.
  - Upload the config to a pastebin such as gist.github.com.
  
  My entire config is in the debug log, but it shouldn't be necessary for this bug, as it only deals with my mouse [buttons] that I have not rebound.

- **Stack Trace:**
  - This is only needed if sway crashes.
  - If you use systemd, you should be able to open the coredump of the most recent crash with gdb with
    `coredumpctl gdb sway` and then `bt full` to obtain the stack trace.
  - If the lines mentioning sway or wlroots have `??` for the location, your binaries were built without debug symbols. Please compile both sway and wlroots from source and try to reproduce.

- **Description:**
  - The steps you took in plain English to reproduce the problem.
  1. open a window (in this case my terminal)
  2. drag the title bar at least slightly
  3. observe that the pointer disappears

This bug makes it significantly harder to rearrange multiple windows using the mouse.

**Comment by sahinf** [comment]
Following your exact steps on the same sway and wlroots versions, I cannot reproduce. Might be something to do with your config. Maybe you can try to reproduce this with a minimal config and upload that here.

**Comment by cplir-c** [comment]
Reproduced with minimal config (no config file)
I symlinked foot to kitty in my PATH, ran it, and dragged the title bar a ways. My mouse pointer disappeared as usual.

**Comment by sahinf** [comment]
You are using void linux right? It's [sway package](https://github.com/void-linux/void-packages/blob/master/srcpkgs/sway/template) seems to use [wlroots0.17.4](https://github.com/void-linux/void-packages/blob/47e54511a09dc12ebefc085fcb6595c1e7b4b341/srcpkgs/wlroots0.17/template#L3C1-L3C15) as it should, but your debug log shows sway is using wlroots0.17.2. I'm curious how that works / happened. Can you somehow get it to use wlroots0.17.4 and try to reproduce? There may have been patches in 17.3 and 17.4 that fixed this.

**Comment by cplir-c** [comment]
```sh
$ sway -d
00:00:00.000 [INFO] [sway/main.c:338] Sway version 1.9
00:00:00.000 [INFO] [sway/main.c:339] wlroots version 0.17.2
00:00:00.000 [INFO] [sway/main.c:120] Linux voidthorn 6.6.54_1 #1 SMP PREEMPT_DYNAMIC
...
```sh
$ xbps-query -l | grep wlroots
ii wlroots0.17-0.17.4_1                                        Modular Wayland compositor library 0.17
ii xdg-desktop-portal-wlr-0.7.1_1                              Backend of xdg-desktop-portal for wlroots
```
Uptime: 2 days
I had tried updating my packages and restarting sway, but I'm still getting the same mouse pointer issue.

Is my package setup broken?
Maybe this is a Void issue at this point.

**Comment by sahinf** [comment]
> Is my package setup broken?
Maybe this is a Void issue at this point.

I'm not familiar with void linux but you can verify this by testing with the master build of sway (1.10) and wlroots (0.19). 

https://github.com/swaywm/sway/wiki/Development-Setup#compiling-as-a-subproject

```
# Clone repositories
git clone https://github.com/swaywm/sway.git
cd sway
git clone https://gitlab.freedesktop.org/wlroots/wlroots.git subprojects/wlroots

# Build sway and wlroots
meson build/
ninja -C build/

# Start sway
build/sway/sway
```

and you should see wlroots 0.19.0 or newer

![image](https://github.com/user-attachments/assets/3eeda40e-dd74-4b1b-8a0c-76506a210c53)

You should be able to test it with a nested instance by launching some window in that nested instance through
`SWAYSOCK=$(find 2>/dev/null /run/user/1000/ -name '*sway*' | tail -n 1) swaymsg 'exec <COMMAND>'`

**Comment by cplir-c** [comment]
I did the compiling...
and it still hid my pointer
but in the log I saw:
```
00:00:36.436 [wlr] [types/wlr_cursor.c:593] XCursor theme is missing 'grab' cursor
```
lol...

I've been opening a subwindow in the nested sway by having ctrl+d to end a terminal in outer sway open my launcher?
Yep, that's what did it. My launcher keybinding is meta+d.

**Comment by wjdp** [comment]
I have the same issue on Ubuntu 24.04 using sway 1.9

**Comment by RunningDroid** [comment]
> You are using void linux right? It's [sway package](https://github.com/void-linux/void-packages/blob/master/srcpkgs/sway/template) seems to use [wlroots0.17.4](https://github.com/void-linux/void-packages/blob/47e54511a09dc12ebefc085fcb6595c1e7b4b341/srcpkgs/wlroots0.17/template#L3C1-L3C15) as it should, but your debug log shows sway is using wlroots0.17.2. I'm curious how that works / happened.

Void packages need to have their `revision` bumped in order to rebuild against updated dependencies, but that should only be needed if the ABI is incompatible afaik

I had the same issue, also because my cursor theme was missing the "grab" icon, so I wrote a script to detect this issue:
```shell
#!/bin/sh
# assume we're running under sway
cursor_theme="$(swaymsg -p -t get_config | grep -Po '^\sxcursor_theme\s\K\w+(?=)')"

# FDO says icons (like cursors) can be in any of:
# $HOME/.icons
# $XDG_DATA_DIRS/icons (typically "/usr/local/share/icons:/usr/share/icons", but flatpak adds more and snap probably does too)
# /usr/share/pixmaps
# https://specifications.freedesktop.org/icon-theme-spec/0.13/#directory_layout
# ArchWiki says to include $HOME/.local/share/icons, probably because of how they interpret $XDG_DATA_DIRS
icon_dirs="$HOME/.local/share/icons:$HOME/.icons"
IFS=':'
if [ ${XDG_DATA_DIRS:-unset} = 'unset' ]; then
	icon_dirs="$icon_dirs:/usr/local/share/icons:/usr/share/icons"
else
	for path in $XDG_DATA_DIRS; do
		icon_dirs="$icon_dirs:$path/icons"
	done
fi
icon_dirs="$icon_dirs:/usr/share/pixmaps"
for path in $icon_dirs; do
	if [ -d "${path}/${cursor_theme}" ] && [ -f "${path}/${cursor_theme}/cursors/grab" ]; then
		printf "Found the grab cursor! it's here: %s\n" "${path}/${cursor_theme}/cursors/grab"
		exit 0
	fi
done
printf 'No grab cursor found for current theme (%s)\n' "$cursor_theme"

# check if the workaround's been applied
for path in $icon_dirs; do
	if [ -d "${path}/default" ] && [ -h "${path}/default/cursors/grab" ]; then
		printf "Found a workaround here: %s\n" "${path}/default/cursors/grab"
	fi
done

```

There's a workaround mentioned [on ArchWiki](https://wiki.archlinux.org/title/Cursor_themes#Create_links_to_missing_cursors) for this kind of issue, but I can't get it to work for some reason. (I probably have to kill all instances of sway.)

**Comment by cplir-c** [comment]
I tried running your script without changes under `dash`, and then `bash`, but neither worked ?
```
user@laptop ~> dash ~/dev/scripts/repairs/foreign-detect-nograb-cursor.sh
/home/user/dev/scripts/repairs/foreign-detect-nograb-cursor.sh: 13: [: /home/user/.local/share/flatpak/exports/share: unexpected operator
No grab cursor found for current theme ()
user@laptop ~> bash ~/dev/scripts/repairs/foreign-detect-nograb-cursor.sh
/home/user/dev/scripts/repairs/foreign-detect-nograb-cursor.sh: line 13: [: too many arguments
No grab cursor found for current theme ()
```

**Comment by cplir-c** [comment]
I've been having the pointer disappear when I move it into an okular window with an open PDF when I've selected an annotation tool as well. I suspect these issues are related?

I tried looking in the places listed on the arch wiki and `~/.local/share/icons` and `~/.icons` both don't exist, but the system-wide directory does. So I ran the symlinking script for each theme with a `cursors` folder:
```sh
#! /bin/sh
for dir in /usr/share/icons/*/cursors ; do
    ln -s right_ptr arrow
    ln -s cross crosshair
    ln -s right_ptr draft_large
    ln -s right_ptr draft_small
    ln -s cross plus
    ln -s left_ptr top_left_arrow
    ln -s cross tcross
    ln -s hand hand1
    ln -s hand hand2
    ln -s left_side left_tee
    ln -s left_ptr ul_angle
    ln -s left_ptr ur_angle
    ln -s left_ptr_watch 08e8e1c95fe2fc01f976f1e063a24ccd
done
```

**Comment by RunningDroid** [comment]
@cplir-c: I apparently altered my local copy to fix shellcheck complaining about missing quotes but forgot to change what I'd put here, it's fixed now

**Comment by cplir-c** [comment]
void linux has updated sway to version 1.11_1 and I'm not experiencing this issue anymore
